File: 24-Graph-Part-II/python/02_dijkstras_algorithm.py
import heapq
from typing import List

class Solution:
    def dijkstra(self, V: int, edges: List[List[int]], S: int) -> List[int]:
        adj = [[] for _ in range(V)]
        for u, v, weight in edges:
            adj[u].append((v, weight))
            adj[v].append((u, weight))

        dist = [float('inf')] * V
        dist[S] = 0

        # (distance, node)
        pq = [(0, S)]

        while pq:
            d, node = heapq.heappop(pq)

            if d > dist[node]:
                continue

            for nxt, weight in adj[node]:
                if dist[node] + weight < dist[nxt]:
                    dist[nxt] = dist[node] + weight
                    heapq.heappush(pq, (dist[nxt], nxt))

        return dist

File: 24-Graph-Part-II/cpp/02-dijkstras-algorithm.cpp
#include <vector>
#include <queue>
#include <climits>

using namespace std;

class Solution {
public:
    vector<int> dijkstra(int V, vector<vector<int>>& edges, int S) {
        vector<vector<pair<int, int>>> adj(V);
        for (auto& edge : edges) {
            int u = edge[0];
            int v = edge[1];
            int weight = edge[2];
            adj[u].push_back({v, weight});
            adj[v].push_back({u, weight});
        }

        vector<int> dist(V, INT_MAX);
        dist[S] = 0;

        // {distance, node}
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        pq.push({0, S});

        while (!pq.empty()) {
            auto [d, node] = pq.top();
            pq.pop();

            if (d > dist[node]) continue;

            for (auto& neighbor : adj[node]) {
                int next = neighbor.first;
                int weight = neighbor.second;
                if (dist[node] + weight < dist[next]) {
                    dist[next] = dist[node] + weight;
                    pq.push({dist[next], next});
                }
            }
        }

        return dist;
    }
};
